Sources/Utility/Process.swift

```diff
@@ -165,6 +171,21 @@ public final class Process: ObjectIdentifierProtocol {
         self.redirectOutput = redirectOutput 
     }
 
+    /// Validate that the program provided exists on the search path and is executable.
+    func validateArgumentIsExecutable(_ program: String) throws {
+
+        // Check to see if the argument is already listed in the vaildated cache.
+        guard !Process.validatedExecutables.contains(program) else {
+            return
+        }
+        let envSearchPaths = Utility.getEnvSearchPaths(pathString: getenv("PATH"), currentWorkingDirectory: currentWorkingDirectory)
```

Comment:
Thanks @aciidb0mb3r, I'll update to work out doing this - and see about adding the caching for the path lookups as well.

So the general mechanism here, if I'm groking it, is to create a DispatchQueue and set it up as a class property, so there's just one of these in the system, and then when we need to run something that leveraging things needing thread safety, to do that relevant work within a call to the DispatchQueue `sync` method, handing in an executable block that returns a value.

I started reading through the guides on DispatchQueues, and the code examples there were in C and talked to creating either Serial or Concurrent quyeyeqs. - I wasn't clear what the initializer in swift returned - does that by default create a serial one? I'll keep reading, but thought I'd ask if you any good pointers to more info here.